### PR TITLE
Feat/53 handle different languages

### PR DESCRIPTION
**Pull request by PiaSchroeder**
Closes https://github.com/CorrelAid/pystatis/issues/53
- used decimal points or commas in pd.read_csv() depending on the chosen language so that floats are read as floats instead of strings
- introduced a database and language specific dictionary for column names so that prettify_table works for en and de

### CODE REVIEW

File: src/pystatis/config.py [hunk 1]
```diff
@@ -32,6 +32,34 @@
...
+            "variable_code": "Auspraegung_Code",
+            "variable_label": "Merkmal_Label",
+            "variable_level": "Auspraegung_Label",
+        },
+        "en": {
+            "time_label": "time_label",
+            "time": "time",
+            "variable_code": "variable_code.1",
+            "variable_label": "variable_label",
+            "variable_level": "variable_code.2",
```
Comment (pmayd) [inline src/pystatis/config.py:49]: is this deliberate that we have `variable_code` here two times, although it is a level? Seems strange to me

File: src/pystatis/table.py [hunk 2]
```diff
@@ -34,10 +36,20 @@ def get_data(self, area: str = "all", prettify: bool = True, **kwargs):
             area (str, optional): Area to search for the object in GENESIS-Online. Defaults to "all".
             prettify (bool, optional): Reformats the table into a readable format. Defaults to True.
         """
-        params = {"name": self.name, "area": area, "format": "ffcsv"}
+        params = {
+            "name": self.name,
+            "area": area,
+            "format": "ffcsv",
+            "language": "de",
```
Comment (pmayd) [inline src/pystatis/table.py:43]: we should probably document that we set de as default language somewhere more prominent, like in the function's doc string but also in the general documentation

File: src/pystatis/table.py [hunk 3]
```diff
@@ -34,10 +36,20 @@ def get_data(self, area: str = "all", prettify: bool = True, **kwargs):
...
+        }
 
         params |= kwargs
 
+        if params["language"] not in ["de", "en"]:
+            raise NotImplementedError(
```
Comment (pmayd) [inline src/pystatis/table.py:49]: this error type is normally not used for cases like this, it is really meant to signal that code needs to be implemented, but here it is more about a `ValueError` because the parameter has the wrong value as it should be one of "de" or "en"

Comment (pmayd) [reply src/pystatis/table.py:49]: maybe we should even implement our own ErrorType here to signal that this is a `QueryParameterError`

File: src/pystatis/table.py [hunk 4]
```diff
@@ -46,13 +58,25 @@ def get_data(self, area: str = "all", prettify: bool = True, **kwargs):
 
         self.raw_data = raw_data_str
         data_buffer = StringIO(raw_data_str)
-        self.data = pd.read_csv(
-            data_buffer, sep=";", na_values=["...", ".", "-", "/", "x"]
-        )
+
+        if params["language"] == "en":
+            self.data = pd.read_csv(
```
Comment (pmayd) [inline src/pystatis/table.py:63]: instead of repeating the whole pd command, set a value for a variable that is used for the decimal, like:

```python
decimal_char = "." if params["language"] == "en" else ","
pd.read_csv(..., decimal=decimal_char)
```

File: src/pystatis/table.py [hunk 5]
```diff
@@ -62,7 +86,7 @@ def get_data(self, area: str = "all", prettify: bool = True, **kwargs):
         self.metadata = metadata
 
     @staticmethod
-    def prettify_table(data: pd.DataFrame, db_name: str) -> pd.DataFrame:
+    def prettify_table(data: pd.DataFrame, db_name: str, language: str) -> pd.DataFrame:
         """Reformat the data into a more readable table
 
         Args:
```
Comment (pmayd) [inline src/pystatis/table.py:92]: you need to update the documentation as well

File: src/pystatis/table.py [hunk 6]
```diff
@@ -75,80 +99,89 @@ def prettify_table(data: pd.DataFrame, db_name: str) -> pd.DataFrame:
...
+        attributes = data.filter(like=column_name_dict["variable_level"])
+        attributes.columns = data.filter(like=column_name_dict["variable_label"]).iloc[-1].tolist()
 
         # Selects all columns containing the values
         values = data.filter(like="__")
 
         # Given a name like BEV036__Bevoelkerung_in_Hauptwohnsitzhaushalten__1000
         # extracts the readable label and omit both the code and the unit
-        values.columns = [name.split("__")[1] for name in values.columns]
+        values.columns = [re.split(r"_{2,}", name)[1] for name in values.columns]
```
Comment (pmayd) [inline src/pystatis/table.py:131]: why was this changed? This is a change of the logic not related to the language, or is it related to the language problem? We should not change core logic without a good reason. Did you encounter a case where there are more than two underscores?

**Review by pmayd — APPROVED** [review]
I left a few comments for small improvements but overall, looks good!

File: src/pystatis/config.py [hunk 7]
```diff
@@ -32,6 +32,34 @@
     "zensus": re.compile(r"^\d{4}[A-Z]-\d{4}$"),
     "regio": re.compile(r"^((\d{5}-.{1,2}($|-.*$))|(A.*$)|([0-9A-Z]{10}$))"),
 }
+COLUMN_NAME_DICT = {
+    "genesis": {
+        "de": {
+            "time_label": "Zeit_Label",
+            "time": "Zeit",
+            "variable_code": "Auspraegung_Code",
```
Comment (pmayd) [inline src/pystatis/config.py:40]: we should stick with the official vocabulary: "Merkmal" is a variable and "Ausprägung" is the value for a variable, so I would expect "Auspraegung_Code" to be "value_code", not "variable_code" etc.

File: src/pystatis/config.py [hunk 8]
```diff
@@ -32,6 +32,34 @@
...
+            "variable_label": "Merkmal_Label",
+            "variable_level": "Auspraegung_Label",
```
Comment (pmayd) [inline src/pystatis/config.py:42]: again, this is a value_label, why level? anyway, should be value, not variable

File: src/pystatis/config.py (hunk 7 above)
Comment (PiaSchroeder) [reply src/pystatis/config.py:40]: As far as I can tell the term "value" is used for the actual numeric value columns and the value_label is the name of that numeric value. "Merkmal" is the variable but "Ausprägung" does not seem to have a consistent English name (or at least I couldn't identify it). In Genesis/Regionalstatistik the respective columns are referred to as variable_label.2 and in Zensus, they seem to be called attributes (variable_attribute_label). I called them "levels" to reflect the (typically nominal) scaling levels of the corresponding variable. So e.g. (variable_label = Bundesland, variable_level = Niedersachsen, value_label = Personen, value = 123). I could change "level" to "attribute" if you think that's more in keeping with the genesis/zensus naming conventions (although I don't think attribute is the right term for what the respective columns actually contain..)?

File: src/pystatis/table.py (hunk 6 above)
Comment (PiaSchroeder) [reply src/pystatis/table.py:131]: Yes, some tables have multiple consecutive underscores (e.g. 81000-0001). With the old version, this led the column names to be empty strings.

File: src/pystatis/config.py (hunk 1 above)
Comment (PiaSchroeder) [reply src/pystatis/config.py:49]: It seems strange to me too, but that's how genesis and Regionalstatistik name their columns... (at least in the tables I tested)

Comment (bergnerjonas) [reply src/pystatis/config.py:48]: There seems to be an issue with the genesis API itself, the response body maps the columns `Merkmal_Code`, `Auspraegung_Code` and `Auspraegung_Label` all to `variable_code` when setting the language to `en`. Our parsing to a pandas df then creates the .1 and .2 suffixes to avoid three duplicated columns.

**Comment by bergnerjonas** [comment]
Will add tests for language support once https://github.com/CorrelAid/pystatis/pull/99 is merged.

File: src/pystatis/config.py (hunk 7 above)
Comment (bergnerjonas) [reply src/pystatis/config.py]: Switched the keys to "value" instead of "variable_level" to be more generic (the variables reflect Months etc. as well).

File: src/pystatis/config.py (hunk 1 above)
Comment (bergnerjonas) [reply src/pystatis/config.py]: Contacted API developers. Will incorporate possible changes in future releases.

**Comment by bergnerjonas** [comment]
Will not support language='en' for Regionalstatistik database, since the API responses are of low data quality and handling them is too complicated and error prone.

File: src/pystatis/table.py [hunk 9]
```diff
@@ -104,88 +125,103 @@ def get_data(
...
         values = data.filter(like="__")
 
         # Given a name like BEV036__Bevoelkerung_in_Hauptwohnsitzhaushalten__1000
-        # extracts the label and the unit and omit the code
-        values.columns = [name.split("__", maxsplit=1)[1] for name in values.columns]
+        # extracts the label and the unit and omits the code
+        values.columns = [re.split(r"_{2,}", name, maxsplit=1)[1] for name in values.columns]
 
-        pretty_data = pd.concat([time, attributes, values], axis=1)
+        pretty_data = pd.concat([time, attributes, values], axis=1).dropna(axis=0, how="all")
```
Comment (MarcoHuebner) [inline src/pystatis/table.py:173]: Not sure whether we actually want dropna, as NaN's contain a lot of "irregular" but relevant information (see other open issues). At first glance, nothing else to complain about :)

Comment (bergnerjonas) [reply src/pystatis/table.py:173]: This only drops rows containing only NaNs

Comment (bergnerjonas) [reply src/pystatis/table.py:173]: This  was added by @PiaSchroeder , I will take a look at the reason for this.

Comment (pmayd) [reply src/pystatis/table.py:173]: should make sense I mean if there is no information at all why keep it but yeah would be nice to document an evas number where this actually happens and why

File: src/pystatis/config.py [hunk 10]
```diff
@@ -33,6 +33,46 @@
...
+        },
+        "en": {
+            "time_label": "time_label",
+            "time": "time",
+            "variable_label": "variable_label",
+            "value_label": "variable_code.2",
+            "value_code": "variable_code.1",
+        },
+    },
+    # Curently, response does not change colum names between languages.
```
Comment (pmayd) [inline src/pystatis/config.py:53]: good comment!

File: src/pystatis/config.py [hunk 11]
```diff
@@ -33,6 +33,46 @@
     "zensus": re.compile(r"^\d{4}[A-Z]-\d{4}$"),
     "regio": re.compile(r"^((\d{5}-.{1,2}($|-.*$))|(A.*$)|([0-9A-Z]{10}$)|(\d{5}\w-Z-\d{1,2}))"),
 }
+COLUMN_NAME_DICT = {
```
Comment (pmayd) [inline src/pystatis/config.py:36]: maybe change into something more speaking like `LANG_TO_COL_MAPPING` but definitely better ;) The name should indicate that the variable holds column names by language or a mapping between the two

File: src/pystatis/db.py [hunk 12]
```diff
@@ -4,12 +4,13 @@
...
 from pystatis.cache import normalize_name
+from pystatis.exception import PystatisConfigError
 
 logger = logging.getLogger(__name__)
 
 
-def identify_db(name: str) -> list[str]:
-    """Identify the required database by matching the item code to the database regex.
+def identify_db_matches(name: str) -> list[str]:
+    """Identify possible databases by matching the item code to the database regex.
```
Comment (pmayd) [inline src/pystatis/db.py:13]: good catch

File: src/pystatis/db.py [hunk 13]
```diff
@@ -28,6 +29,35 @@ def identify_db(name: str) -> list[str]:
     return db_match
 
 
+def identify_db(table_name: str) -> str:
```
Comment (pmayd) [inline src/pystatis/db.py:32]: this method should be in db.py not in table.py in my opinion. Didn't we had this function before?

File: src/pystatis/db.py [hunk 14]
```diff
@@ -28,6 +29,35 @@ def identify_db(name: str) -> list[str]:
...
+    exist, use the first with existing credentials.
+    In this case, it must be a Cube (provided all regexing works as intended).
+
+    Args:
+        table_name (str): Query parameter 'name' corresponding to the item code.
+
+    Returns:
+        db_name (str): Identified database.
+    """
+    db_match = identify_db_matches(table_name)
```
Comment (pmayd) [inline src/pystatis/db.py:44]: should be plural then

File: src/pystatis/db.py [hunk 15]
```diff
@@ -28,6 +29,35 @@ def identify_db(name: str) -> list[str]:
...
+
+    if db_match:
+        for db_name in db_match:
+            # Return first hit with existing credentials.
+            if check_credentials(db_name):
+                return db_name
+        raise PystatisConfigError(
```
Comment (pmayd) [inline src/pystatis/db.py:51]: why do we need this check in this function? This function should only return a database given a string, but it actually checks if the credentials are valid...so it does not really identify the database. I thought this check is done in a different place, isn't it? I don't think we need it here?

Comment (pmayd) [reply src/pystatis/db.py:51]: and the `raise` command could be put in a `else` branch which makes it even clearer that this code is only execute when the for loop ends without any match

File: src/pystatis/http_helper.py [hunk 16]
```diff
@@ -99,22 +100,7 @@ def get_data_from_endpoint(endpoint: str, method: str, params: dict, db_name: st
         name = params.get("name", params.get("selection", ""))
 
         if name is not None:
-            db_match = db.identify_db(name)
```
Comment (pmayd) [inline src/pystatis/http_helper.py:102]: ah here it is! I think this is now a matter of opinion but `identify_db` should not be responsible for checking credentials as this depends on the provided user information and not on the passed string but I can accept both views as you could argue that identify_db can only identify a valid database if there are valid credentials...Normally I try to keep low level functions as small as possible and I think it is common and good practice to add additional checks like here with the credentials in the caller function so they can decide what makes a valid response and what not. But if you have good arguments to move these checks into the indentify_db function, I will align

File: src/pystatis/table.py (hunk 9 above)
Comment (MarcoHuebner) [reply src/pystatis/table.py:173]: The behaviour is IMO problematic, as all values with some sort of quality sign (or other issues) would get removed - and from my limited experience, this behaviour is quite common. I would therefore leave removing NaNs up the the user and would propose to just assess proper handling of the NaNs.

File: src/pystatis/http_helper.py [hunk 17]
```diff
@@ -99,22 +100,7 @@ def get_data_from_endpoint(endpoint: str, method: str, params: dict, db_name: st
...
-                    if db.check_credentials(name):
-                        db_name = name
-                        break
-                else:
-                    raise PystatisConfigError(
-                        "Missing credentials!\n"
-                        f"To access this item you need to be a registered user of: {db_match} \n"
-                        "Please run setup_credentials()."
-                    )
+            db_name = identify_db(name)
```
Comment (pmayd) [inline src/pystatis/http_helper.py:103]: anyway, this function can now raise two different exceptions, `PystatisConfigError` and `ValueError` and I think we should except them here, maybe

File: src/pystatis/db.py [hunk 18]
```diff
@@ -28,6 +29,35 @@ def identify_db(name: str) -> list[str]:
...
+def identify_db(table_name: str) -> str:
+    """Identify database by matching with the provided item code.
+    This is done by matching the item code to the database regex. If multiple matches
+    exist, use the first with existing credentials.
+    In this case, it must be a Cube (provided all regexing works as intended).
+
+    Args:
+        table_name (str): Query parameter 'name' corresponding to the item code.
+
+    Returns:
```
Comment (pmayd) [inline src/pystatis/db.py:41]: Pls add a "Raises" section explaining the two exceptions being thrown

File: src/pystatis/http_helper.py [hunk 19]
```diff
@@ -99,22 +100,7 @@ def get_data_from_endpoint(endpoint: str, method: str, params: dict, db_name: st
...
 
     if not db_name:
```
Comment (pmayd) [inline src/pystatis/http_helper.py:105]: If I am not mistaken, your changes to `identify_db` make this case impossible as `db_name` always exists at this point as otherwise there would have been an exception. So we either remove this check here or we use a try except blocvk around identify_db and in case of any exception we set `db_name=None` or empty string...but does not make much sense to catch an exception just to throw it again...so I think we can safely remove it from here, correct?

File: src/pystatis/table.py [hunk 20]
```diff
@@ -92,10 +102,21 @@ def get_data(
         raw_data_header = raw_data_lines[0]
         raw_data_str = raw_data_header + "".join(line for line in raw_data_lines[1:] if line[:4].isdigit())
         data_buffer = StringIO(raw_data_str)
-        self.data = pd.read_csv(data_buffer, sep=";", na_values=["...", ".", "-", "/", "x"])
+
+        decimal_char = "," if language == "de" else "."
```
Comment (pmayd) [inline src/pystatis/table.py:106]: if it is not reused, we can put this directly as argument to `pd.read_csv`

File: src/pystatis/table.py [hunk 21]
```diff
@@ -104,88 +125,103 @@ def get_data(
...
+            language (str): The requested language. One of "de" or "en".
 
         Returns:
             pd.DataFrame: Formatted dataframe that omits all unnecessary Code columns
             and includes informative columns names
         """
         match db_name:
             case "genesis":
-                pretty_data = Table.parse_genesis_table(data)
+                pretty_data = Table.parse_genesis_table(data, language)
```
Comment (pmayd) [inline src/pystatis/table.py:142]: If `language` is used throughout the class, maybe we should add it as parameter to `__init__` and make a proper class instance variable out of it? Default can still be "de"

File: tests/conftest.py [hunk 22]
```diff
@@ -13,16 +11,13 @@ def vcr_config():
...
-    params = args[1]
-
-    if "table_name" in args[0]:
-        param = [param for param in params if param[0] in name][0]
-        table_name = param[0]
-        return table_name
-
-    return name
+    """Name of the VCR cassette, changed to {language}_{table_name}."""
+    test_arguments = request.node.callspec.params
```
Comment (pmayd) [inline tests/conftest.py:15]: why did you need to change this? The way it should work, to my understanding, is that all requests are added to the same {table_name}.yaml. If you request the same table name with "de" and then with "en" these are different language parameters, so the two requests should be added to the same .yml file in my previous version. I would prefer this over twice the amount of yaml files. And it is actually better to have both versions in one file for comparison reasons

**Review by pmayd — CHANGES_REQUESTED** [review]
A few minor changes I would like to request and a discussion about the changed `vcr_casette_name` fixture. Overall, great work with the tests

File: src/pystatis/table.py (hunk 9 above)
Comment (pmayd) [reply src/pystatis/table.py:173]: But why are quality signs removed when dropping all NANs? If there is a quality sign then the row cannot be only NaNs...

Comment (MarcoHuebner) [reply src/pystatis/table.py:173]: Sorry, was not reading carefully enough and overlooked "all". Resolving this conversation

File: src/pystatis/db.py (hunk 15 above)
Comment (bergnerjonas) [reply src/pystatis/db.py:51]: The logic is the following:

1. Find all possible DBs, given the table name.
2. Go through all DB candidates and check if the user provided credentials for a candidate. If so: return that as the matched DB

File: src/pystatis/table.py [hunk 23]
```diff
@@ -81,7 +85,13 @@ def get_data(
             "format": "ffcsv",
         }
 
-        raw_data_bytes = load_data(endpoint="data", method="tablefile", params=params)
+        db_name = db.identify_db(self.name)
+        if db_name == "regio" and language != "de":
+            raise QueryParameterError(f"Language '{language}' is not supported for Regionalstatistik, only 'de'.")
+        elif language not in ["de", "en"]:
+            raise QueryParameterError(f"Language '{language}' is not supported. Please choose from: ['de', 'en'].")
```
[the remark this answers is not in the document]
Comment (bergnerjonas) [reply src/pystatis/table.py:92]: It's really only meant to check that we do not support EN for regionalstatistik, even though the API theoretically supports it.

File: tests/conftest.py (hunk 22 above)
Comment (bergnerjonas) [reply tests/conftest.py:15]: Good point, I thought this would overwrite existing request. I will test this and will rollback the changes.

File: src/pystatis/http_helper.py (hunk 19 above)
Comment (bergnerjonas) [reply src/pystatis/http_helper.py]: Good catch!

File: src/pystatis/http_helper.py (hunk 16 above)
Comment (bergnerjonas) [reply src/pystatis/http_helper.py:102]: I split up the functionality into to separate methods now.

File: src/pystatis/table.py (hunk 21 above)
Comment (bergnerjonas) [reply src/pystatis/table.py:142]: I am not sure about this. In general, I would favor treating the language just like any other query parameter and pass it to get_data.

Comment (bergnerjonas) [reply src/pystatis/table.py:142]: However, it might make the code easier if we treat it as a class instance variable...

File: src/pystatis/table.py [hunk 24]
```diff
@@ -104,88 +125,103 @@ def get_data(
...
+    def parse_genesis_table(data: pd.DataFrame, language: str) -> pd.DataFrame:
         """Parse GENESIS table ffcsv format into a more readable format"""
-        # Extracts time column with name from first element of Zeit_Label column
-        time = pd.DataFrame({data["Zeit_Label"].iloc[0]: data["Zeit"]})
 
-        # Extracts new column names from first values of the Merkmal_Label columns
-        # and assigns these to the relevant attribute columns (Auspraegung_Label)
-        attributes = data.filter(like="Auspraegung_Label")
-        attributes.columns = data.filter(like="Merkmal_Label").iloc[0].tolist()
+        column_name_dict = COLUMN_NAME_DICT["genesis-regio"][language]
```
[the remark this answers is not in the document]
Comment (bergnerjonas) [reply src/pystatis/table.py]: Not sure about this - often we only need the value of a key from the column_name_dict once and adding a dedicated variable for this seems like quite the overhead for me.

[the remark this answers is not in the document]
Comment (pmayd) [reply src/pystatis/table.py]: That's true but as I said it is hard to maintain such code when the dict is all over the place. Imagine we want to change the dict structure, then you have to visit all the places the dict is used. Defining all access variables at the beginning makes it much easier to maintain and understand how it is used ;) But up to you

File: src/pystatis/table.py (hunk 21 above)
Comment (pmayd) [reply src/pystatis/table.py:142]: I don't think many people will use "en" anyway so let's keep it simple right now and everything as parameter of get_data so we can provide a single interface

**Review by pmayd — APPROVED** [review]
Good job once you are finished with everything

File: tests/conftest.py (hunk 22 above)
Comment (bergnerjonas) [reply tests/conftest.py:15]: Indeed this works as you suggested. I rolled back my changes to separate the cassettes by language.

File: src/pystatis/table.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (bergnerjonas) [reply src/pystatis/table.py]: Fair enough - I extracted them at the beginning of the method.
